backend/doi/invenio.go

```diff
@@ -0,0 +1,190 @@
+func resolveInvenioEndpoint(ctx context.Context, srv *rest.Client, pacer *fs.Pacer, resolvedURL *url.URL) (provider Provider, endpoint *url.URL, err error) {
+	fs.Logf(nil, "invenioURL = %s", resolvedURL.String())
+
+	var res *http.Response
+	opts := rest.Opts{
+		Method:  "GET",
+		RootURL: resolvedURL.String(),
+	}
+	err = pacer.Call(func() (bool, error) {
+		res, err = srv.Call(ctx, &opts)
+		return shouldRetry(ctx, res, err)
+	})
+	if err != nil {
+		return "", nil, err
+	}
+
+	// First, attempt to grab the API URL from the headers
+	var linksetURL *url.URL
+	links := parseLinkHeader(res.Header.Get("Link"))
+	for _, link := range links {
+		if link.Rel == "linkset" && link.Type == "application/linkset+json" {
+			parsed, err := url.Parse(link.Href)
+			if err == nil {
+				linksetURL = parsed
+				break
+			}
+		}
+	}
+
+	if linksetURL != nil {
+		endpoint, err = checkInvenioAPIURL(ctx, srv, pacer, linksetURL)
+		if err == nil {
+			return Invenio, endpoint, nil
+		}
+		fs.Logf(nil, "using linkset URL failed: %s", err.Error())
+	}
+
+	// If there is no linkset header, try to grab the record ID from the URL
+	recordID := ""
+	resURL := res.Request.URL
+	fs.Logf(nil, "resURL = %s", resURL.String())
+	match := invenioRecordRegex.FindStringSubmatch(resURL.EscapedPath())
+	if match != nil {
+		recordID = match[1]
+		guessedURL := res.Request.URL.ResolveReference(&url.URL{
+			Path: "/api/records/" + recordID,
+		})
+		endpoint, err = checkInvenioAPIURL(ctx, srv, pacer, guessedURL)
+		if err == nil {
+			return Invenio, endpoint, nil
+		}
+		fs.Logf(nil, "guessing the URL failed: %s", err.Error())
+	}
+
+	return "", nil, fmt.Errorf("could not resolve the Invenio API endpoint for '%s'", resolvedURL.String())
+}
+
+func checkInvenioAPIURL(ctx context.Context, srv *rest.Client, pacer *fs.Pacer, resolvedURL *url.URL) (endpoint *url.URL, err error) {
+	var result api.InvenioRecordResponse
+	opts := rest.Opts{
+		Method:  "GET",
+		RootURL: resolvedURL.String(),
+	}
+	err = pacer.Call(func() (bool, error) {
+		res, err := srv.CallJSON(ctx, &opts, nil, &result)
+		return shouldRetry(ctx, res, err)
+	})
+	if err != nil {
+		return nil, err
+	}
+	if result.Links.Self == "" {
+		return nil, fmt.Errorf("could not parse API response from '%s'", resolvedURL.String())
+	}
+	return url.Parse(result.Links.Self)
+}
+
+// Implements Fs.List() for Invenio
+func (f *Fs) listInvenio(ctx context.Context, dir string) (entries fs.DirEntries, err error) {
+	if dir != "" {
+		return nil, fs.ErrorDirNotFound
+	}
+
+	fileEntries, err := f.listInvevioDoiFiles(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error listing %q: %w", dir, err)
+	}
+	for _, entry := range fileEntries {
+		entries = append(entries, entry)
+	}
+	return entries, nil
+}
+
+// List the files contained in the DOI
+func (f *Fs) listInvevioDoiFiles(ctx context.Context) (entries []*Object, err error) {
+	// Use the cache if populated
+	cachedEntries, found := f.cache.GetMaybe("files")
```

Comment:
question: Do we need to worry about accessing stuff that is too stale from the cache? Do we have some way to invalidate the cache? Or is this unnecessary and once we have something in the cache we can be certain that it does not need to be updated.

Reply:
> Or is this unnecessary and once we have something in the cache we can be certain that it does not need to be updated.
Correct, the file list from a published dataset should not change. It may change if a new version is published or if file hosting is updated.

Reply:
So if we can just put some simple TTL on the cache it may be a good idea. Because it is not like the dataset can NEVER change. But again we can just wait on the review from the rclone authors here.

Reply:
Otherwise, it will be fine in most reasonable cases.